### ISSUE TITLE
Self quantification in data declarations

### ISSUE THREAD

**Issue by RafaelBocquet** [goldfirere/ghc#56]
Currently, this declaration :

``` haskell
data A :: * where B :: forall (x :: A). A
```

results in a "AThing evaluated unexpectedly" ghc panic.

I think it should be accepted.

**Comment by goldfirere** [comment]
Yes, it should be accepted. But arranging for it to be accepted is a little challenging (purely from an engineering point of view), and might not make it for 8.0. We'll get there eventually. :)

The panic will go away though.

**Comment by RafaelBocquet** [comment]
ghc now hangs when compiling this code.
